include/pybind11/detail/class.h

```diff
@@ -223,7 +223,7 @@ inline bool deregister_instance_impl(void *ptr, instance *self) {
     auto &registered_instances = get_internals().registered_instances;
     auto range = registered_instances.equal_range(ptr);
     for (auto it = range.first; it != range.second; ++it) {
-        if (Py_TYPE(self) == Py_TYPE(it->second)) {
+        if (Py_TYPE(self) == Py_TYPE(it->second) && self == it->second) {
```

Comment:
If `self == it->second`, wouldn't `Py_TYPE(self) == Py_TYPE(it->second)` be a useless comparison?

(`Py_TYPE` is `#define Py_TYPE(ob)             (((PyObject*)(ob))->ob_type)` in CPython's headers)

Reply:
And given this, it's important to figure out why `Py_TYPE` was used in the first place?

I do think [this concern](https://github.com/pybind/pybind11/issues/1568#issuecomment-435686117) was adequately addressed by [this reply](https://github.com/pybind/pybind11/issues/1568#issuecomment-435701894).